includes/Modules/AdSense/Ad_Blocking_Recovery_Tag.php

```diff
@@ -0,0 +1,114 @@
+<?php
+/**
+ * Class Google\Site_Kit\Modules\AdSense\Ad_Blocking_Recovery_Tag
+ *
+ * @package   Google\Site_Kit\Modules\AdSense
+ * @copyright 2023 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Modules\AdSense;
+
+use Google\Site_Kit\Core\Storage\Setting;
+
+/**
+ * Class for AdSense Ad Blocker Recovery Tag.
+ *
+ * @since n.e.x.t
+ * @access private
+ * @ignore
+ */
+class Ad_Blocking_Recovery_Tag extends Setting {
+	const OPTION = 'googlesitekit_adsense_ad_blocking_recovery_tag';
+
+	/**
+	 * Gets ad blocking recovery tag.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return array Array with tag and error protection code.
+	 */
+	public function get() {
+		$option = parent::get();
+
+		$tag                   = base64_decode( $option['tag'], true );
+		$error_protection_code = base64_decode( $option['error_protection_code'], true );
```

Comment:
Basically what we do for Ad Blocking Recovery can be easily considered as XSS since we inject a 3rd party script on frontend pages. We must be extremely careful here to make sure that what we get from the option is not a malicious code that a hacker managed to save into the database.

Base64 encoding is a very bad choice for encryption here because anyone can use it to create a malicious script and save it into the database that we will inject on each page. We must use the `Data_Encryption` class to really encrypt these values. Probably the best option for this class would be to override constructor to wrap passed `$options` with a new instance of the `Encrypted_Options` class and then pass the new encrypted options to the parent constructor.

cc @aaemnnosttv

Reply:
My perspective is that in order for someone/something to inject code into the string we output, they already need to have access to the database somehow, because the tag is fetched and saved on the server, not from user input. So there needs to be:

* a man-in-the-middle between Site Kit’s server and Google, OR
* database access

for an attacker to write to this string.

At that point, there are LOTS of other vectors for users to inject `<script>` tags, etc. They can place it in `post content`, etc. I suppose the only thing we’re doing here is making sure that our echo statement isn’t the one outputting bad content, but if a user has enough access to read encryption keys and database records, etc. they can manage to replace the data we store.

Ultimately someone has to have so much access to mess with that data that it’s game over anyway, so it feels like we’re trying to prevent anything bad from happening in Site Kit when an attacker can already do way worse things. So the added complexity of encryption, hashing, etc. for data that needs to be read on every page render and needs to be decrypted seemed a bit misguided and complex for very limited benefit.

The base64 handling wasn’t to encrypt so much as just prevent any automated scripts/plugins that went through database records from messing with anything that looks like a `<script>` tag. I could even see misguided “security” plugins searching for these strings inside DB records and thinking they ARE hacking attempts, and removing them, so the base64 approach was just about obscuring things enough to prevent silly overwrites.

But ultimately, the nature of this tag—because it’s always-changing from AdSense and thus essentially needs to be output as a raw, unverifiable/checkable string, means that we have limited options.

Hashing it and verifying that hash before rendering (or using the encryption library) would prevent someone from messing with that field to output malicious HTML, but there are other places that are better spots to output that HTML (eg. `post_content`).

I don’t mean to hand-wave away security, but in order for this to even be an attractive attack vector the site would need:

1. Site Kit
2. AdSense
3. Ad Blocking Recovery
4. A plugin/database vulnerability that allows arbitrary write access to database fields (which functionality is already game-over for their site)
5. A security plugin that restricts post_content from outputting <script> tags thus making that an unviable attack vector but that also doesn’t restrict other HTML output from plugins

I think by the time we’re in trouble here, it’s already game-over for that site. If another plugin wants to write arbitrary SQL to a website it can.

It doesn't _really_ matter if a bad actor writes a BitCoin miner `<script>` tag into the HTML _we_ output or `post_content` outputs. I suppose the main concern would be a vulnerability in our plugin that allows users to write specifically to that option from the client-side. If that's the concern, then I suppose we do need to encrypt it. But otherwise I think the end result is inescapable when there is the ability for _any other plugin_ to write _any content to the database_. 🤷🏻‍♂️

Reply:
> I suppose the main concern would be a vulnerability in our plugin that allows users to write specifically to that option from the client-side. If that's the concern, then I suppose we do need to encrypt it.

@tofumatt @aaemnnosttv, yes, but not in our plugin. That vulnerability can be in any 3rd party plugin or theme that would let a bad guy save options with an arbitrary name from the client side (for example, poorly secured custom REST API endpoint). In this case, they can use that plugin/theme to save a malicious script with our option name and it will be rendered by our plugin on the frontend -> we will be guilty for XSS injections. If we use `Encrypted_Options`, then they won't be able to use this approach to fake our script.

> Also, if we do decide to use encryption, this would not require changes to `Ad_Blocking_Recovery_Tag` since it accepts an `Options_Interface` as its constructor dependency, so we could encrypt the entire value just by passing an instance of `Encrypted_Options` instead.

Yes, this is kind of the same what I suggested. It can be done either way: in the class that instantiates an instance of the `Ad_Blocking_Recovery_Tag` class or in the constructor of the class itself if we still want to pass just `$options` from the AdSense class (that doesn't have an instance of encrypted options and needs to create one too).

Reply:
Fair enough! I suppose it's relatively harmless for us to encrypt that data to prevent that attack vector. Thanks for catching this one @eugene-manuilov 👍🏻